# HG changeset patch
# Parent 620d826bce9527620cfa726d5df0e4d9af5ddf3b
add 'C' flag to 'guioptions' which control to show the caption of window

diff --git a/runtime/doc/options.txt b/runtime/doc/options.txt
index 2378aa9..34e7a1f 100644
--- a/runtime/doc/options.txt
+++ b/runtime/doc/options.txt
@@ -3826,6 +3826,9 @@ A jump table for the options with a short description can be found at |Q_op|.
 	  'F'	Add a footer.  Only for Motif.  See |gui-footer|.
 
 
+	  'C'	Remove caption (title) bar.  Support Win32 only.
+
+
 						*'guipty'* *'noguipty'*
 'guipty'		boolean	(default on)
 			global
diff --git a/src/gui.c b/src/gui.c
index f7ec508..27d386e 100644
--- a/src/gui.c
+++ b/src/gui.c
@@ -3336,6 +3336,8 @@ gui_init_which_components(char_u *oldval UNUSED)
     static int	prev_tearoff = -1;
     int		using_tearoff = FALSE;
 #endif
+    static int	prev_nocaption = -1;
+    int		using_caption = TRUE;
 
     char_u	*p;
     int		i;
@@ -3413,6 +3415,9 @@ gui_init_which_components(char_u *oldval UNUSED)
 		using_tearoff = TRUE;
 #endif
 		break;
+	    case GO_NOCAPTION:
+		using_caption = FALSE;
+		break;
 	    default:
 		/* Ignore options that are not supported */
 		break;
@@ -3523,6 +3528,14 @@ gui_init_which_components(char_u *oldval UNUSED)
 	    prev_tearoff = using_tearoff;
 	}
 #endif
+	if (using_caption != prev_nocaption)
+	{
+#if defined(WIN3264)
+	    gui_mch_show_caption(using_caption);
+#endif
+	    prev_nocaption = using_caption;
+	    need_set_size = TRUE;
+	}
 	if (need_set_size != 0)
 	{
 #ifdef FEAT_GUI_GTK
diff --git a/src/gui_w32.c b/src/gui_w32.c
index a21bb4b..250719e 100644
--- a/src/gui_w32.c
+++ b/src/gui_w32.c
@@ -40,6 +40,9 @@ static int s_directx_load_attempted = 0;
 static int gui_mswin_get_menu_height(int fix_window);
 #endif
 
+static int get_caption_height(void);
+static int get_caption_width_adjustment(void);
+
 #if defined(FEAT_DIRECTX) || defined(PROTO)
     int
 directx_enabled(void)
@@ -3138,11 +3141,12 @@ gui_mswin_get_valid_dimensions(
 
     base_width = gui_get_base_width()
 	+ (GetSystemMetrics(SM_CXFRAME) +
-	   GetSystemMetrics(SM_CXPADDEDBORDER)) * 2;
+	   GetSystemMetrics(SM_CXPADDEDBORDER)) * 2
+	+ get_caption_width_adjustment();
     base_height = gui_get_base_height()
 	+ (GetSystemMetrics(SM_CYFRAME) +
 	   GetSystemMetrics(SM_CXPADDEDBORDER)) * 2
-	+ GetSystemMetrics(SM_CYCAPTION)
+	+ get_caption_height()
 #ifdef FEAT_MENU
 	+ gui_mswin_get_menu_height(FALSE)
 #endif
@@ -3537,9 +3541,11 @@ gui_mch_newfont(void)
     if (win_socket_id == 0)
     {
 	gui_resize_shell(rect.right - rect.left
+	    - get_caption_width_adjustment()
 	    - (GetSystemMetrics(SM_CXFRAME) +
 	       GetSystemMetrics(SM_CXPADDEDBORDER)) * 2,
 	    rect.bottom - rect.top
+	    - get_caption_height()
 	    - (GetSystemMetrics(SM_CYFRAME) +
 	       GetSystemMetrics(SM_CXPADDEDBORDER)) * 2
 	    - GetSystemMetrics(SM_CYCAPTION)
@@ -4675,6 +4681,57 @@ gui_is_win32s(void)
     return (os_version.dwPlatformId == VER_PLATFORM_WIN32s);
 }
 
+    static int
+has_caption(void)
+{
+    return GetWindowLong(s_hwnd, GWL_STYLE) & WS_CAPTION;
+}
+
+    static int
+get_caption_height(void)
+{
+    /*
+     * A window's caption includes extra 1 dot margin.  When caption is
+     * removed the margin also be removed.  So we must return -1 when
+     * caption is diabled.
+     */
+    return has_caption() ? GetSystemMetrics(SM_CYCAPTION) : -1;
+}
+
+    static int
+get_caption_width_adjustment(void)
+{
+    return has_caption() ? 0 : -2;
+}
+
+    void
+gui_mch_show_caption(int show)
+{
+    const static LONG flags_on = WS_CAPTION;
+    const static LONG flags_off = 0;
+    LONG style, newstyle;
+
+    /* Remove caption when title is null. */
+    style = newstyle = GetWindowLong(s_hwnd, GWL_STYLE);
+    if (show)
+    {
+	newstyle &= ~flags_off;
+	newstyle |= flags_on;
+    }
+    else
+    {
+	newstyle &= ~flags_on;
+	newstyle |= flags_off;
+    }
+    if (newstyle != style)
+    {
+	SetWindowLong(s_hwnd, GWL_STYLE, newstyle);
+	SetWindowPos(s_hwnd, NULL, 0, 0, 0, 0,
+		SWP_FRAMECHANGED | SWP_NOMOVE | SWP_NOSIZE | SWP_NOZORDER);
+	gui_set_shellsize(FALSE, FALSE, RESIZE_BOTH);
+    }
+}
+
 #ifdef FEAT_MENU
 /*
  * Figure out how high the menu bar is at the moment.
@@ -4714,7 +4771,7 @@ gui_mswin_get_menu_height(
 	    {
 		RECT r1, r2;
 		int frameht = GetSystemMetrics(SM_CYFRAME);
-		int capht = GetSystemMetrics(SM_CYCAPTION);
+		int capht = get_caption_height();
 
 		/* get window rect of s_hwnd
 		 * get client rect of s_hwnd
@@ -5954,10 +6011,11 @@ gui_mch_set_shellsize(int width, int height,
 
     /* compute the size of the outside of the window */
     win_width = width + (GetSystemMetrics(SM_CXFRAME) +
-			 GetSystemMetrics(SM_CXPADDEDBORDER)) * 2;
+			 GetSystemMetrics(SM_CXPADDEDBORDER)) * 2
+			+ get_caption_width_adjustment();
     win_height = height + (GetSystemMetrics(SM_CYFRAME) +
 			   GetSystemMetrics(SM_CXPADDEDBORDER)) * 2
-			+ GetSystemMetrics(SM_CYCAPTION)
+			+ get_caption_height()
 #ifdef FEAT_MENU
 			+ gui_mswin_get_menu_height(FALSE)
 #endif
@@ -6864,7 +6922,8 @@ gui_mch_get_screen_dimensions(int *screen_w, int *screen_h)
 
     *screen_w = workarea_rect.right - workarea_rect.left
 		- (GetSystemMetrics(SM_CXFRAME) +
-		   GetSystemMetrics(SM_CXPADDEDBORDER)) * 2;
+		   GetSystemMetrics(SM_CXPADDEDBORDER)) * 2
+		- get_caption_width_adjustment();
 
     /* FIXME: dirty trick: Because the gui_get_base_height() doesn't include
      * the menubar for MSwin, we subtract it from the screen height, so that
@@ -6872,7 +6931,7 @@ gui_mch_get_screen_dimensions(int *screen_w, int *screen_h)
     *screen_h = workarea_rect.bottom - workarea_rect.top
 		- (GetSystemMetrics(SM_CYFRAME) +
 		   GetSystemMetrics(SM_CXPADDEDBORDER)) * 2
-		- GetSystemMetrics(SM_CYCAPTION)
+		- get_caption_height()
 #ifdef FEAT_MENU
 		- gui_mswin_get_menu_height(FALSE)
 #endif
diff --git a/src/option.h b/src/option.h
index 38d4caf..2669169 100644
--- a/src/option.h
+++ b/src/option.h
@@ -238,7 +238,8 @@
 #define GO_TOOLBAR	'T'		/* add toolbar */
 #define GO_FOOTER	'F'		/* add footer */
 #define GO_VERTICAL	'v'		/* arrange dialog buttons vertically */
-#define GO_ALL		"aAbcefFghilmMprtTv" /* all possible flags for 'go' */
+#define GO_NOCAPTION	'C'		/* remove caption bar */
+#define GO_ALL		"aAbcCefFghilmMprtTv" /* all possible flags for 'go' */
 
 /* flags for 'comments' option */
 #define COM_NEST	'n'		/* comments strings nest */
diff --git a/src/proto/gui_w32.pro b/src/proto/gui_w32.pro
index 0c31f9c..6ac0dc6 100644
--- a/src/proto/gui_w32.pro
+++ b/src/proto/gui_w32.pro
@@ -61,6 +61,7 @@ char_u *gui_mch_browsedir(char_u *title, char_u *initdir);
 char_u *gui_mch_browse(int saving, char_u *title, char_u *dflt, char_u *ext, char_u *initdir, char_u *filter);
 int get_cmd_args(char *prog, char *cmdline, char ***argvp, char **tofree);
 int gui_is_win32s(void);
+void gui_mch_show_caption(int show);
 void gui_mch_set_parent(char *title);
 void gui_mch_prepare(int *argc, char **argv);
 int gui_mch_init(void);
